src/MongoDB.Driver.Core/Core/Authentication/ScramSha1Authenticator.cs

```diff
@@ -36,294 +37,43 @@ public sealed class ScramSha1Authenticator : SaslAuthenticator
         /// Initializes a new instance of the <see cref="ScramSha1Authenticator"/> class.
         /// </summary>
         /// <param name="credential">The credential.</param>
         public ScramSha1Authenticator(UsernamePasswordCredential credential)
-            : this(credential, new DefaultRandomStringGenerator())
-        {
-        }
+            : this(credential, new DefaultRandomStringGenerator()) { }
 
         internal ScramSha1Authenticator(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
-            : base(new ScramSha1Mechanism(credential, randomStringGenerator))
-        {
-            _databaseName = credential.Source;
-        }
+            : base(credential, HashAlgorithmName.SHA1, randomStringGenerator, H1, Hi1, Hmac1) { }
 
-        // properties
-        /// <inheritdoc/>
-        public override string DatabaseName
+        private static byte[] H1(byte[] data)
         {
-            get { return _databaseName; }
-        }
-
-        // nested classes
-        private class ScramSha1Mechanism : ISaslMechanism
-        {
-            private readonly UsernamePasswordCredential _credential;
-            private readonly IRandomStringGenerator _randomStringGenerator;
-
-            public ScramSha1Mechanism(UsernamePasswordCredential credential, IRandomStringGenerator randomStringGenerator)
-            {
-                _credential = Ensure.IsNotNull(credential, nameof(credential));
-                _randomStringGenerator = Ensure.IsNotNull(randomStringGenerator, nameof(randomStringGenerator));
-            }
-
-            public string Name
-            {
-                get { return MechanismName; }
-            }
-
-            public ISaslStep Initialize(IConnection connection, ConnectionDescription description)
-            {
-                Ensure.IsNotNull(connection, nameof(connection));
-                Ensure.IsNotNull(description, nameof(description));
-
-                const string gs2Header = "n,,";
-                var username = "n=" + PrepUsername(_credential.Username);
-                var r = GenerateRandomString();
-                var nonce = "r=" + r;
-
-                var clientFirstMessageBare = username + "," + nonce;
-                var clientFirstMessage = gs2Header + clientFirstMessageBare;
-
-                return new ClientFirst(
-                    Utf8Encodings.Strict.GetBytes(clientFirstMessage),
-                    clientFirstMessageBare,
-                    _credential,
-                    r);
-            }
-
-            private string GenerateRandomString()
-            {
-                const string legalCharacters = "!\"#$%&'()*+-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~";
-
-                return _randomStringGenerator.Generate(20, legalCharacters);
-            }
-
-            private static string PrepUsername(string username)
+            using (var sha1 = SHA1.Create())
```

Comment:
Most of this functionality got moved to the base class (ScramShaAuthenticator). The key responsibilities of derived classes are now to supply the H, Hi, and HMAC functions as described in the RFC.